File: provider-ci/src/generate.ts
import { Config } from "./config";
import { generateExamplesFiles } from "./examples";
import { generateProviderFiles } from "./provider";

export interface RepositoryFile {
  path: string;
  data: unknown;
}

export interface RepositoryFiles {
  repository: string;
  files: RepositoryFile[];
}

export function generate(config: Config): RepositoryFiles {
  switch (config.template) {
    case "bridged":
      return {
        repository: `pulumi-${config.provider}`,
        files: generateProviderFiles(config),
      };
    case "examples":
      return {
        repository: "examples",
        files: generateExamplesFiles(),
      };
  }
}

File: package-ci/internal/cmd/generate.go
package cmd

import (
	"github.com/pulumi/ci-mgmt/package-ci/internal/pkg"
	"github.com/spf13/cobra"
)

var generateOpts = &pkg.GenerateOpts{}

// generateCmd represents the generate command
var generateCmd = &cobra.Command{
	Use:   "generate",
	Short: "Generate repository files.",
	RunE: func(cmd *cobra.Command, args []string) error {
		err := pkg.GeneratePackage(pkg.GenerateOpts{
			RepositoryName: generateOpts.RepositoryName,
			OutDir:         generateOpts.OutDir,
			TemplateName:   generateOpts.TemplateName,
			ConfigPath:     generateOpts.ConfigPath,
		})
		return err
	},
}

func init() {
	rootCmd.AddCommand(generateCmd)

	generateCmd.Flags().StringVarP(&generateOpts.RepositoryName, "name", "n", "", "repository name to generate, e.g.: pulumi/pulumi-aws")
	generateCmd.MarkFlagRequired("name")

	generateCmd.Flags().StringVarP(&generateOpts.OutDir, "out", "o", ".", "directory to generate files to")
	generateCmd.MarkFlagRequired("out")

	/*
		type GenerateOpts struct {
			PackageName  string // e.g.: pulumi-aws
			OutDir       string
			TemplateName string // path inside templates, e.g.: bridged-provider
			ConfigPath   string // .yaml file containing template config
		}
	*/

	generateCmd.Flags().StringVarP(&generateOpts.TemplateName, "template", "t", "", "template name to use, e.g.: bridged-provider")
	generateCmd.MarkFlagRequired("template")
	generateCmd.Flags().StringVarP(&generateOpts.ConfigPath, "config", "c", "", "config file to use, e.g.: config.yaml")
	generateCmd.MarkFlagRequired("config")
}
